### PR TITLE
[2.7] bpo-36337: Use ssize_t and size_t for socket.send()/sendall()

### PR DESCRIPTION
**Pull request by matrixise**
This PR is not yet completed (missing tests), I have "backported" a patch from 3.x.

https://bugs.python.org/issue36337

### CODE REVIEW

**Comment by vstinner** [comment]
I don't think that a test is needed: I cannot find a test on master.

Your PR is wrong on Windows. Look at the master branch:

```
#ifdef MS_WINDOWS
    if (ctx->len > INT_MAX)
        ctx->len = INT_MAX;
    ctx->result = send(s->sock_fd, ctx->buf, (int)ctx->len, ctx->flags);
#else
    ctx->result = send(s->sock_fd, ctx->buf, ctx->len, ctx->flags);
#endif
```

On Windows, the type of send() length argument is int, not size_t.

File: Modules/socketmodule.c [hunk 1]
```diff
@@ -2797,7 +2797,9 @@ static PyObject *
 sock_send(PySocketSockObject *s, PyObject *args)
 {
     char *buf;
-    int len, n = -1, flags = 0, timeout;
+    int flags = 0, timeout;
+    ssize_t n = -1;
```
Comment (kmiku7) [inline Modules/socketmodule.c]: I think using `Py_ssize_t` is more portable. It's defined [here](https://github.com/python/cpython/blob/2.7/Include/pyport.h#L182).
`Py_buffer` also use this to declare its fields.

File: Modules/socketmodule.c [hunk 2]
```diff
@@ -2797,7 +2797,8 @@ static PyObject *
 sock_send(PySocketSockObject *s, PyObject *args)
 {
     char *buf;
-    int len, n = -1, flags = 0, timeout;
+    int flags = 0, timeout;
+    Py_ssize_t len, n = -1;
```
Comment (vstinner) [inline Modules/socketmodule.c:2801]: return PyInt_FromLong((long)n); below is wrong

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by matrixise** [comment]
I have made the requested changes; please review again

File: Modules/socketmodule.c [hunk 3]
```diff
@@ -2813,12 +2814,18 @@ sock_send(PySocketSockObject *s, PyObject *args)
     BEGIN_SELECT_LOOP(s)
     Py_BEGIN_ALLOW_THREADS
     timeout = internal_select_ex(s, 1, interval);
-    if (!timeout)
+    if (!timeout) {
 #ifdef __VMS
         n = sendsegmented(s->sock_fd, buf, len, flags);
```
Comment (vstinner) [inline Modules/socketmodule.c:2819]: That's wrong: sendsegmented() len type is int, not Py_ssize_t. Don't forget the VMS! :-)

File: Misc/NEWS.d/next/Library/2019-03-18-10-08-30.bpo-36337.QhJnXy.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Use ssize_t and size_t instead of int for the send() function (see
+http://man7.org/linux/man-pages/man2/send.2.html (Linux) and
+https://nixdoc.net/man-pages/FreeBSD/man2/send.2.html (FreeBSD)).
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2019-03-18-10-08-30.bpo-36337.QhJnXy.rst]: I propose:

Fix buffer overflow in :meth:`~socket.socket.send` and :meth:`~socket.socket.sendall` methods of :func:`socket.socket` for data larger than 2 GiB.

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by matrixise** [comment]
I have made the requested changes; please review again

File: Modules/socketmodule.c [hunk 5]
```diff
@@ -2813,12 +2814,18 @@ sock_send(PySocketSockObject *s, PyObject *args)
     BEGIN_SELECT_LOOP(s)
     Py_BEGIN_ALLOW_THREADS
     timeout = internal_select_ex(s, 1, interval);
-    if (!timeout)
+    if (!timeout) {
 #ifdef __VMS
-        n = sendsegmented(s->sock_fd, buf, len, flags);
+        n = sendsegmented(s->sock_fd, buf, (int)len, flags);
```
Comment (vstinner) [inline Modules/socketmodule.c]: (int)len is unsafe if len is larger than INT_MAX. You should modify sendsegmented() to accept Py_ssize_t rather than int.

Comment (matrixise) [reply Modules/socketmodule.c]: :+1: good catch and I was thinking about this problem.

**Review by vstinner — APPROVED** [review]
LGTM.

The subtle side effect of this change is that send() and sendall() now always return a Python 'long' object, whereas they always returned a Python 'int' object previously. I think that it's acceptable, but it might be risky to push so change late in 2.7 lifetime (EOL scheduled at the beginning of next year).

On the other side, currently, calling sendall() with more than 2 GiB data loose data. But well, who does that with Python 2.7?

I would prefer to have the review of another core dev.

@benjaminp @serhiy-storchaka @pablogsal: Would you mind to review this change? Do you think that it's ok to change send/sendall return type (int => long, supposed to be compatible)?

File: Modules/socketmodule.c [hunk 6]
```diff
@@ -2830,7 +2837,7 @@ sock_send(PySocketSockObject *s, PyObject *args)
     PyBuffer_Release(&pbuf);
     if (n < 0)
         return s->errorhandler();
-    return PyInt_FromLong((long)n);
+    return PyLong_FromSsize_t(n);
```
Comment (serhiy-storchaka) [inline Modules/socketmodule.c]: Just use `PyInt_FromSsize_t()`.

**Comment by matrixise** [comment]
I have made the requested changes; please review again

@serhiy-storchaka @vstinner

File: Modules/socketmodule.c [hunk 7]
```diff
@@ -625,16 +625,16 @@ set_gaierror(int error)
...
 {
     int n = 0;
-    int remaining = len;
+    Py_ssize_t remaining = len;
 
     while (remaining > 0) {
         unsigned int segment;
 
-        segment = (remaining >= SEGMENT_SIZE ? SEGMENT_SIZE : remaining);
+        segment = ((size_t)remaining >= SEGMENT_SIZE ? SEGMENT_SIZE : (unsigned int) remaining);
```
Comment (pablogsal) [inline Modules/socketmodule.c:637]: Wait, is this cast safe? In C99 ssize_t can be an unsigned long, unless I am missing something. I think you need:

```
Py_SAFE_DOWNCAST(remaining, Py_ssize_t, unsigned int)
```

Comment (vstinner) [reply Modules/socketmodule.c:637]: SEGMENT_SIZE is 32 KiB: remaining is only casted for values between [0; 32 KiB]:

```
#define SEGMENT_SIZE (32 * 1024 -1)
```

**Review by pablogsal — APPROVED** [review]
LGTM

**Pull request merged**